packages/message-box/src/index.vue

```diff
@@ -0,0 +1,369 @@
+  setup(props) {
+    let vm
+    const popup = usePopup(props, doClose)
+    const state = reactive<State>({
+      uid: 1,
+      title: undefined,
+      message: '',
+      type: '',
+      iconClass: '',
+      customClass: '',
+      showInput: false,
+      inputValue: null,
+      inputPlaceholder: '',
+      inputType: 'text',
+      inputPattern: null,
+      inputValidator: null,
+      inputErrorMessage: '',
+      showConfirmButton: true,
+      showCancelButton: false,
+      action: '',
+      confirmButtonText: '',
+      cancelButtonText: '',
+      confirmButtonLoading: false,
+      cancelButtonLoading: false,
+      confirmButtonClass: '',
+      confirmButtonDisabled: false,
+      cancelButtonClass: '',
+      editorErrorMessage: null,
+      callback: null,
+      dangerouslyUseHTMLString: false,
+      focusAfterClosed: null,
+      isOnComposition: false,
+      distinguishCancelAndClose: false,
+      type$: '',
+    })
+    const icon = computed(() => state.iconClass || (state.type && TypeMap[state.type] ? `el-icon-${ TypeMap[state.type] }` : ''))
+
+    const confirmButtonClasses = computed(() => `el-button--primary ${ state.confirmButtonClass }`)
+
+    const cancelButtonClasses = computed(() => `${ state.cancelButtonClass }`)
+
+    watch(() => state.inputValue, val => {
+      nextTick().then(() => {
+        if (state.type$ === 'prompt' && val !== null) {
+          validate()
+        }
+      })
+    }, { immediate: true })
+
+    watch(() => popup.state.visible, val => {
+      if (val) {
+        state.uid++
+        if (state.type$ === 'alert' || state.type$ === 'confirm') {
+          nextTick().then(() => { vm.refs.confirm.$el.focus() })
+        }
+        state.focusAfterClosed = document.activeElement
+        dialog = new Dialog(vm.vnode.el, state.focusAfterClosed, getFirstFocus())
+      }
+      if (state.type$ !== 'prompt') return
+      if (val) {
+        setTimeout(() => {
+          if (vm.refs.input && vm.refs.input.$el) {
+            getInputElement().focus()
+          }}, 500)
+      } else {
+        state.editorErrorMessage = ''
+        removeClass(getInputElement(), 'invalid')
+      }
+    })
+
+    onMounted(async () => {
+      vm = getCurrentInstance()
+      await nextTick()
+      if (props.closeOnHashChange) {
+        window.addEventListener('hashchange', popup.close)
+      }
+    })
+
+    onBeforeUnmount(() => {
+      if (props.closeOnHashChange) {
+        window.removeEventListener('hashchange', popup.close)
+      }
+      setTimeout(() => {
+        dialog.closeDialog()
+      })
+    })
+
+    function getSafeClose () {
+      const currentId = state.uid
+      return () => {
+        nextTick().then(() => {
+          if (currentId === state.uid) doClose()
+        })
+      }
+    }
+
+    function doClose() {
+      if (!popup.state.visible) return
+      popup.state.visible = false
+      popup.updateClosingFlag(true)
+
+      vm.ctx.onClose && vm.ctx.onClose()
```

Comment:
`ctx` may be replaced with `props` which is not `excluded`.

`props` is a proxy property of `vnode.component`, component will be updated when modify `props` outside.